File: repos/openrappter/python/openrappter/agents/google_voice_agent.py
"""Google Voice watch decisions, on the grail bones.

This is the Python half of `typescript/src/telephony/watch.ts`. The two are
deliberately transliterations of each other and are held together by a shared
fixture, `tests/google-voice-parity.json`, which both sides assert against.

WHY THE DECISIONS ARE SHARED BUT THE TRANSPORT IS NOT
-----------------------------------------------------
openrappter can drive Chrome over the DevTools Protocol, so on a laptop the
Google Voice rung really can send and receive. A grail brainstem running under
Pyodide or Azure Functions has no browser, no subprocess, and no socket to give
it one — so on those tiers the rung simply is not available, and this agent says
so rather than pretending.

What must NOT differ is the judgement. If the same inbox produces a reply on one
platform and silence on the other, then "the twin runs on both bones" is not
true, and which machine happened to wake up first becomes a behavioural fact.
So every rule below is byte-identical in intent to the TypeScript, and the
parity fixture fails loudly if the two ever drift.

ARTICLE VII
-----------
Nothing here imports anything outside the standard library, touches the network,
reads a clock, or shells out. `now` is always a parameter. That is what makes
the file portable to every tier, and it is enforced by an AST test rather than
by good intentions.

THE FAILURE THIS EXISTS TO PREVENT
----------------------------------
An always-on agent that can send texts fails dangerously in one direction only:
replying to things it should not, at machine speed, to real people. First run
against years of history. Answering its own outbound. Two bots ping-ponging.
Re-answering the same message every poll because state was lost. Every rule is
one of those, and the default answer is NO.
"""

import json

from agents.basic_agent import BasicAgent

# Every agent in the repo has to declare one of these (conformance R2/R3) so a
# strain can govern it without reading the source. This file had none.
__manifest__ = {
    "schema": "rapp-agent/1.0",
    "name": "@openrappter/google-voice",
    "version": "1.0.0",
    "display_name": "Google Voice",
    "description": (
        "Decides whether an inbound text may be answered — rate limits, self-reply "
        "and loop guards, quiet hours — and returns the decision. It never sends."
    ),
    "author": "Kody Wildfeuer",
    "ring": "ga",
    # Empty, and provably so: this file imports `json` and the agent base class
    # and nothing else. It is the decision core described at the top of this
    # module — "no browser, no subprocess, and no socket to give" — so the
    # sending, the polling and the state are all somebody else's capability to
    # declare. The TypeScript GoogleVoiceAgent is the half that touches the
    # world, and it declares network and filesystem-write.
    "capabilities": [],
    "tags": ["openrappter", "google-voice", "telephony", "safety"],
    "category": "communication",
    "quality_tier": "official",
    "requires_env": [],
}

DEFAULT_POLICY = {
    # Four replies an hour to one number is plenty for a negotiation and nowhere
    # near enough to be a nuisance if something goes wrong.
    "maxRepliesPerThread": 4,
    "windowMs": 60 * 60 * 1000,
    # A message older than a day is history, not a live conversation.
    "maxAgeMs": 24 * 60 * 60 * 1000,
}


def empty_state():
    """The state a watcher starts from."""
    return {"knownThreads": {}, "handled": [], "replies": {}}


def _digits(value):
    out = "".join(ch for ch in (value or "") if ch.isdigit())
    if len(out) == 11 and out.startswith("1"):
        out = out[1:]
    return out


def _same_number(a, b):
    da, db = _digits(a), _digits(b)
    return bool(da) and da == db


def decide(message, state, policy, now):
    """Whether the watcher may act on one message.

    `now` is a parameter, never a clock read, so the same inputs always produce
    the same verdict — which is what allows the two implementations to be
    compared against a fixture at all.
    """
    # We do not talk to ourselves. An outbound bubble is our own voice, and
    # treating it as input is how a loop starts.
    if message.get("direction") == "outbound":
        return {"act": False, "reason": "outbound", "detail": "our own message"}
    if _same_number(message.get("from"), policy.get("selfNumber")):
        return {"act": False, "reason": "self", "detail": "from our own number"}

    if message.get("id") in state.get("handled", []):
        return {
            "act": False,
            "reason": "already-handled",
            "detail": "already acted on this message",
        }

    # FIRST SIGHT OF A THREAD IS NEVER ACTIONABLE.
    # On first run the inbox is full of history. Answering it would mean texting
    # everyone who has ever messaged this number. The first poll only records a
    # watermark; a thread becomes live from the next message onward.
    known = state.get("knownThreads", {}).get(message.get("threadId"))
    if not known:
        return {
            "act": False,
            "reason": "thread-unseen",
            "detail": "first sight of this thread — recording a watermark instead of replying",
        }
    if message.get("at", 0) <= known.get("watermark", 0):
        return {
            "act": False,
            "reason": "older-than-watermark",
            "detail": "predates the watermark",
        }

    if now - message.get("at", 0) > policy.get("maxAgeMs", DEFAULT_POLICY["maxAgeMs"]):
        return {"act": False, "reason": "too-old", "detail": "older than the freshness window"}

    allow = policy.get("allowFrom")
    if allow:
        if not any(_same_number(n, message.get("from")) for n in allow):
            return {
                "act": False,
                "reason": "not-allowed",
                "detail": "sender is not on the allow list",
            }

    # The loop guard. If something upstream goes wrong — a bot on the other end,
    # a bad prompt, a bug — this stops it at four instead of four hundred.
    window = policy.get("windowMs", DEFAULT_POLICY["windowMs"])
    cap = policy.get("maxRepliesPerThread", DEFAULT_POLICY["maxRepliesPerThread"])
    recent = [t for t in state.get("replies", {}).get(message.get("threadId"), []) if now - t < window]
    if len(recent) >= cap:
        return {
            "act": False,
            "reason": "rate-limited",
            "detail": "already replied %d times to this thread in the window" % len(recent),
        }

    return {"act": True, "reason": "new-inbound", "detail": "new inbound message"}


def observe(state, thread_id, at):
    """Record that a thread was observed. Safe to call every poll."""
    threads = dict(state.get("knownThreads", {}))
    known = threads.get(thread_id)
    threads[thread_id] = {"watermark": max(known.get("watermark", 0), at) if known else at}
    out = dict(state)
    out["knownThreads"] = threads
    return out


def record_reply(state, message, at):
    """Record that we acted on a message and replied to its thread."""
    replies = dict(state.get("replies", {}))
    replies[message.get("threadId")] = list(replies.get(message.get("threadId"), [])) + [at]
    # Keep `handled` bounded so a long-lived daemon's state cannot grow without
    # limit; the watermark already covers anything older.
    handled = (list(state.get("handled", [])) + [message.get("id")])[-500:]
    threads = dict(state.get("knownThreads", {}))
    prev = threads.get(message.get("threadId"), {}).get("watermark", 0)
    threads[message.get("threadId")] = {"watermark": max(prev, message.get("at", 0))}
    out = dict(state)
    out["handled"] = handled
    out["replies"] = replies
    out["knownThreads"] = threads
    return out


class GoogleVoiceAgent(BasicAgent):
    """Decide what an always-on watcher may do about Google Voice messages."""

    def __init__(self):
        super().__init__(
            name="GoogleVoice",
            metadata={
                "name": "GoogleVoice",
                "description": (
                    "Decides whether an always-on watcher may reply to a Google Voice "
                    "message. Shares its judgement byte-for-byte with openrappter so the "
                    "same inbox behaves the same on either set of bones. Decisions only — "
                    "sending requires a browser this tier may not have."
                ),
                "parameters": {
                    "type": "object",
                    "properties": {
                        "action": {
                            "type": "string",
                            "description": "decide | observe | record | policy",
                        },
                        "message": {"type": "object", "description": "The inbox message."},
                        "state": {"type": "object", "description": "Watcher state."},
                        "policy": {"type": "object", "description": "Overrides for DEFAULT_POLICY."},
                        "now": {"type": "number", "description": "Epoch ms. Never read from a clock."},
                    },
                    "required": ["action"],
                },
            },
        )

    def perform(self, **kwargs):
        action = kwargs.get("action", "decide")
        state = kwargs.get("state") or empty_state()
        policy = dict(DEFAULT_POLICY)
        policy.update(kwargs.get("policy") or {})
        message = kwargs.get("message") or {}
        now = kwargs.get("now")

        if action == "policy":
            return json.dumps({"status": "ok", "policy": policy}, indent=2)

        if action == "observe":
            return json.dumps(
                {"status": "ok", "state": observe(state, message.get("threadId"), message.get("at", 0))},
                indent=2,
            )

        if action == "record":
            return json.dumps(
                {"status": "ok", "state": record_reply(state, message, now or message.get("at", 0))},
                indent=2,
            )

        if now is None:
            # Refusing is the honest move: this tier may have no trustworthy
            # clock, and silently substituting one would make the verdict
            # unreproducible and break parity with the other bones.
            return json.dumps(
                {
                    "status": "error",
                    "error": "`now` (epoch ms) is required — this agent never reads a clock.",
                },
                indent=2,
            )

        return json.dumps({"status": "ok", "verdict": decide(message, state, policy, now)}, indent=2)

File: repos/openrappter/typescript/src/agents/GoogleVoiceAgent.ts
/**
 * GoogleVoiceAgent — the phone layer as something openrappter can schedule.
 *
 * The launchd installer shipped earlier runs a dedicated always-on process. This
 * is the other shape of the same capability: an ordinary agent that performs ONE
 * poll when invoked, so openrappter's own cron can wake it up. The organism
 * schedules itself instead of asking the operating system to babysit a daemon,
 * which is the difference between a program that happens to run on your machine
 * and one that lives there.
 *
 * WHY A TICK AND NOT A LOOP
 *
 * A cron job that never returns is a cron job that runs once. `perform()` does a
 * single pass and reports what happened; the schedule owns the repetition. That
 * also makes the whole thing testable — one tick is a function call, where a
 * loop is a thing you have to wait for and then kill.
 *
 * WHAT IT REFUSES TO DECIDE
 *
 * Nothing here judges whether a message deserves a reply. That belongs to
 * `telephony/watch.ts`, which is shared byte-for-byte with the grail brainstem's
 * `google_voice_agent.py` and pinned by tests/google-voice-parity.json. If this
 * file started making its own choices, the two platforms would drift and which
 * machine woke up first would become a behavioural fact.
 *
 * Above all it inherits the rule that makes an unattended poll safe at all:
 * first sight of a thread never replies. It records a watermark and says
 * nothing, so scheduling this against an inbox with history does not text
 * everyone who has ever messaged the number.
 */

import { BasicAgent } from './BasicAgent.js';
import type { AgentMetadata } from './types.js';
import { GoogleVoiceWatcher, loadState, STATE_PATH } from '../telephony/watcher.js';
import type { InboxMessage } from '../telephony/watch.js';
import {
  GREETING, ThreadMemory, createAssistantResponder, isAutomated, toSms,
} from '../telephony/reply.js';
import type { LLMProvider } from '../providers/types.js';

/**
 * Conformance R2/R3: every agent declares a manifest so a strain can govern it
 * without reading the source. This file had none.
 */
export const __manifest__ = {
  schema: 'rapp-agent/1.0',
  name: '@openrappter/google-voice',
  version: '1.0.0',
  display_name: 'Google Voice',
  description:
    'Watches a Google Voice inbox and answers texts under the reply policy — '
    + 'rate limits, self-reply and loop guards, quiet hours.',
  author: 'Kody Wildfeuer',
  ring: 'ga',
  // Declared from what this actually reaches through `../telephony/watcher.js`:
  //   network           — drives Google Voice in a Chrome CDP session
  //   filesystem-write  — persists poll state so a restart cannot re-answer
  //                       years of history
  // The decision half is the Python google_voice_agent, which reaches neither
  // and declares nothing.
  capabilities: [
    'network',
    'filesystem-write',
  ],
  tags: [
    'openrappter',
    'google-voice',
    'telephony',
  ],
  category: 'communication',
  quality_tier: 'official',
  requires_env: [],
} as const;

export interface GoogleVoiceAgentOptions {
  /** Answers an inbound message. Returning null means "say nothing". */
  respond?: (message: InboxMessage) => Promise<string | null>;
  port?: number;
  statePath?: string;
  /**
   * Where to get a model. Defaults to the same backend selection the daemon
   * uses, so the agent answers with whatever rung is actually working.
   */
  provider?: LLMProvider | (() => Promise<LLMProvider | null>);
  /** Live agent roster, so "what can you do" is answered from fact. */
  capabilities?: string[];
  /** Remembers the thread across ticks — each cron wake is a fresh process. */
  memory?: ThreadMemory;
}

/**
 * The reply of last resort.
 *
 * This is what the agent says when there is no model to reach: the greeting,
 * once, which is true. For a long time it was also the answer to *everything* —
 * a real thread went "What else" → greeting → "Okay well list 10 things you can
 * do" → "You there?" → the same greeting again. The message was never read.
 *
 * It still refuses automated senders before anything else, because a
 * verification code is not a conversation and quoting one back puts a security
 * code in a thread it was never meant to leave.
 */
export async function defaultResponder(message: InboxMessage): Promise<string | null> {
  if (isAutomated(message.text || '')) return null;
  return toSms(GREETING);
}

/**
 * The responder the agent actually uses: reads the message, answers it, and
 * falls back to `defaultResponder`'s greeting only when no model can be reached.
 *
 * Kept lazy — resolving a backend costs a token exchange or a CLI probe, and a
 * tick that finds an empty inbox should not pay for one.
 */
async function defaultProvider(): Promise<LLMProvider | null> {
  const { selectBackend } = await import('../providers/backend-select.js');
  const {
    hasAuthProfileAuthority,
    resolveGithubToken,
  } = await import('../copilot-check.js');
  const profileAuthority = hasAuthProfileAuthority();
  const githubToken = await resolveGithubToken();
  const choice = await selectBackend({
    githubToken: githubToken ?? undefined,
    allowIndependentCli: !profileAuthority,
    allowAmbientCredentials: !profileAuthority,
  });
  return choice.provider;
}

export class GoogleVoiceAgent extends BasicAgent {
  private readonly options: GoogleVoiceAgentOptions;
  /**
   * Thread memory lives on the agent, not inside a tick. A cron wake-up is a
   * fresh call but the same process, and without this a follow-up like
   * "What else" arrives with no idea what it is following.
   */
  private readonly memory: ThreadMemory;

  constructor(options: GoogleVoiceAgentOptions = {}) {
    const metadata: AgentMetadata = {
      name: 'GoogleVoice',
      description:
        'Check Google Voice for new messages and reply to the ones that deserve it. '
        + 'Performs one poll per invocation so it can be driven by cron. Never replies '
        + 'to a thread it is seeing for the first time, so scheduling it against an '
        + 'existing inbox does not answer its history.',
      parameters: {
        type: 'object',
        properties: {
          action: {
            type: 'string',
            description: 'check (default) — one poll; status — report without polling',
          },
          dryRun: {
            type: 'boolean',
            description: 'Decide and report, but send nothing. The safe way to schedule it first.',
          },
        },
        required: [],
      },
    };
    super('GoogleVoice', metadata);
    this.options = options;
    this.memory = options.memory ?? new ThreadMemory();
  }

  /**
   * Read the message and answer it. Only falls back to the greeting when there
   * is no model to reach — which is a real condition, not the normal path.
   */
  private responder(): (m: InboxMessage) => Promise<string | null> {
    if (this.options.respond) return this.options.respond;
    return createAssistantResponder({
      provider: this.options.provider ?? defaultProvider,
      capabilities: this.options.capabilities,
      memory: this.memory,
    });
  }

  async perform(kwargs: Record<string, unknown>): Promise<string> {
    const action = (kwargs.action as string) ?? 'check';
    const statePath = this.options.statePath ?? STATE_PATH;

    if (action === 'status') {
      const state = await loadState(statePath);
      return JSON.stringify({
        status: 'success',
        knownThreads: Object.keys(state.knownThreads).length,
        handled: state.handled.length,
        message:
          'Threads already seen are live; anything new to this watcher gets a watermark '
          + 'on its first poll rather than a reply.',
      }, null, 2);
    }

    const lines: string[] = [];
    const watcher = new GoogleVoiceWatcher({
      port: this.options.port,
      statePath,
      dryRun: kwargs.dryRun === true,
      respond: this.options.respond ?? this.responder(),
      log: (line) => lines.push(line),
    });

    // A cron tick must load the durable state itself. Without this every wake-up
    // would start from an empty watcher, see every thread as unseen, and — while
    // the first-sight rule keeps that from texting anyone — it would never get
    // past recording watermarks, so the agent would appear to run forever
    // without ever answering anything.
    (watcher as unknown as { state: unknown }).state = await loadState(statePath);

    let replied = 0;
    let error: string | undefined;
    try {
      replied = await watcher.tick();
    } catch (e) {
      error = (e as Error).message;
    }

    const after = await loadState(statePath);
    return JSON.stringify({
      status: error ? 'error' : 'success',
      replied,
      knownThreads: Object.keys(after.knownThreads).length,
      handled: after.handled.length,
      ...(error ? { error } : {}),
      log: lines,
      data_slush: {
        replied,
        known_threads: Object.keys(after.knownThreads).length,
        transport_available: !lines.some((l) => l.includes('no Chrome DevTools endpoint')),
      },
    }, null, 2);
  }
}
